Make Gossip DNS reject gossip messages if clockskew > tombstoneTimeout / 2

tomwilkie opened the pull request:
Fixes #1156 

5mins max clock skew seems too small to me; I suggest we make it bigger (and therefore make the DNS tombstone timeout larger).

bboreham commented on nameserver/nameserver.go:
```diff
@@ -208,12 +214,16 @@ func (n *Nameserver) OnGossipUnicast(sender router.PeerName, msg []byte) error {
 
-	if err := entries.check(); err != nil {
+	if gossip.Timestamp < now()-int64(gossipWindow/time.Second) {
+		return nil, nil, fmt.Errorf("Refusing old gossip message: %d", gossip.Timestamp)
```
Think it would be worth telling the user the apparent difference in time; it's quite hard from a 64-bit int to see whether your clock is just a bit slow or your machine is set to 1970.

bboreham commented on nameserver/nameserver.go:
```diff
@@ -210,12 +218,16 @@ func (n *Nameserver) OnGossipUnicast(sender router.PeerName, msg []byte) error {
 
-	if err := entries.check(); err != nil {
+	if delta := gossip.Timestamp - now(); delta > gossipWindow || delta < -gossipWindow {
+		return nil, nil, fmt.Errorf("Refusing old gossip message (delta %d): %d", delta, gossip.Timestamp)
```
This will get printed out as "connection shutting down due to error: Refusing old gossip message ...", so it doesn't really explain what happened.

How about "clock skew of %v detected" ?

bboreham commented:
Apart from one nit this LOKTM

tomwilkie commented:
@bboreham I've addressed your comment and after discussion with @rade increase the tombstone timeout on DNS to 30mins (allow a max clockskew of 15mins)

tomwilkie commented:
Test failure is flakey TestNameservers so can safely be ignored.  I'm looking into it.
